File: assembler.py
import enum, re, sys
from typing import TextIO, List, IO


class ArgType(enum.Enum):
    V = re.compile('^v0*([0-9a-f])$', re.IGNORECASE)
    V0 = re.compile('^(V0)$', re.IGNORECASE)
    B = re.compile('^(B)$', re.IGNORECASE)
    F = re.compile('^(F)$', re.IGNORECASE)
    DT = re.compile('^(DT)$', re.IGNORECASE)
    ST = re.compile('^(ST)$', re.IGNORECASE)
    I = re.compile('^(I)$', re.IGNORECASE)
    K = re.compile('^(K)$', re.IGNORECASE)
    AI = re.compile('^(\\[I])$', re.IGNORECASE)
    NIBBLE = re.compile('^(?:0x|#)0*([0-9a-f])$', re.IGNORECASE)
    BYTE = re.compile('^(?:0x|#)0*([0-9a-f]|[0-9a-f]{2})$', re.IGNORECASE)
    ADDR = re.compile('^(?:0x|#)0*([0-9a-f]|[0-9a-f]{2}|[0-9a-f]{3})$', re.IGNORECASE)


def get_formats(opcode: str) -> dict:
    formats = {
        'ADD': (
            {'format': [ArgType.I, ArgType.V],
             'assemble': lambda args: 0xF01E | (args[2] << 8)},
            {'format': [ArgType.V, ArgType.BYTE],
             'assemble': lambda args: 0x7000 | (args[1] << 8) | (args[2])},
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x8004 | (args[1] << 8) | (args[2] << 4)},),
        'AND': (
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x8002 | (args[1] << 8) | (args[2] << 4)},),
        'CALL': (
            {'format':[ArgType.ADDR],
             'assemble': lambda args: 0x2000 | args[1]},),
        'CLS': (
            {'format': [],
             'assemble': lambda args: 0x00E0},),
        'DRW': (
            {'format': [ArgType.V, ArgType.V, ArgType.NIBBLE],
             'assemble': lambda args: 0xD000 | (args[1] << 8) | (args[2] << 4) | args[3]},),
        'JP': (
            {'format': [ArgType.ADDR],
             'assemble': lambda args: 0x1000 | args[1]},
            {'format': [ArgType.V0, ArgType.ADDR],
             'assemble': lambda args: 0xB000 | args[2]},),
        'LD': (
            {'format': [ArgType.B, ArgType.V],
             'assemble': lambda args: 0xF033 | (args[2] << 8)},
            {'format': [ArgType.DT, ArgType.V],
             'assemble': lambda args: 0xF015 | (args[2] << 8)},
            {'format': [ArgType.F, ArgType.V],
             'assemble': lambda args: 0xF029 | (args[2] << 8)},
            {'format': [ArgType.I, ArgType.ADDR],
             'assemble': lambda args: 0xA000 | args[2]},
            {'format': [ArgType.AI, ArgType.V],
             'assemble': lambda args: 0xF055 | (args[2] << 8)},
            {'format': [ArgType.ST, ArgType.V],
             'assemble': lambda args: 0xF018 | (args[2] << 8)},
            {'format': [ArgType.V, ArgType.BYTE],
             'assemble': lambda args: 0x6000 | (args[1] << 8) | args[2]},
            {'format': [ArgType.V, ArgType.DT],
             'assemble': lambda args: 0xF007 | (args[1] << 8)},
            {'format': [ArgType.V, ArgType.AI],
             'assemble': lambda args: 0xF065 | (args[1] << 8)},
            {'format': [ArgType.V, ArgType.K],
             'assemble': lambda args: 0xF00A | (args[1] << 8)},
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x8000 | (args[1] << 8) | (args[2] << 4)},),
        'OR': (
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x8001 | (args[1] << 8) | (args[2] << 4)},),
        'RET': (
            {'format': [],
             'assemble': lambda args: 0x00EE},),
        'RND': (
            {'format': [ArgType.V, ArgType.BYTE],
             'assemble': lambda args: 0xC000 | (args[1] << 8) | args[2]},),
        'SE': (
            {'format': [ArgType.V, ArgType.BYTE],
             'assemble': lambda args: 0x3000 | (args[1] << 8) | args[2]},
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x5000 | (args[1] << 8) | (args[2] << 4)},),
        'SHL': (
            {'format': [ArgType.V],
             'assemble': lambda args: 0x800E | (args[1] << 8)},
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x800E | (args[1] << 8) | args[2] << 4}),
        'SHR': (
            {'format': [ArgType.V],
             'assemble': lambda args: 0x8006 | (args[1] << 8)},
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x8006 | (args[1] << 8) | args[2] << 4}),
        'SKNP': (
            {'format': [ArgType.V],
             'assemble': lambda args: 0xE0A1 | (args[1] << 8)},),
        'SKP': (
            {'format': [ArgType.V],
             'assemble': lambda args: 0xE09E | (args[1] << 8)},),
        'SNE': (
            {'format': [ArgType.V, ArgType.BYTE],
             'assemble': lambda args: 0x4000 | (args[1] << 8) | args[2]},
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x9000 | (args[1] << 8) | (args[2] << 4)},),
        'SUB': (
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x8005 | (args[1] << 8) | (args[2] << 4)},),
        'SUBN': (
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x8007 | (args[1] << 8) | (args[2] << 4)},),
        'SYS': (
            {'format': [ArgType.ADDR],
             'assemble': lambda args: args[1]},),
        'XOR': (
            {'format': [ArgType.V, ArgType.V],
             'assemble': lambda args: 0x8003 | (args[1] << 8) | (args[2] << 4)},),
    }
    if opcode.upper().strip() not in formats:
        raise ValueError('Unrecognized opcode: ' + opcode)
    else:
        return formats[opcode.upper().strip()]


def parseMnemonic(line: str, *expected_args: ArgType) -> list:
    retval = []
    opcode, _, args = line.partition(' ')
    args = [a for a in args.split(',') if a]

    if len(args) != len(expected_args):
        retval = None
    else:
        retval.append(opcode)

        for expected_arg in expected_args:
            arg = args.pop(0).strip()
            match = expected_arg.value.match(arg)
            if match is not None:
                try:
                    retval.append(int(match.group(1), 16))
                except ValueError:
                    retval.append(match.group(1))
            else:
                retval = None
                break

    return retval


def split_label(source: str) -> tuple:
    label, colon, instruction = source.partition(':')
    if label and colon:
        if label[0].isalpha() and label.isalnum():
            return label.strip(), instruction.strip()
        else:
            raise ValueError('Invalid label: ' + source)
    else:
        return None, source


def get_define(source: str) -> tuple:
    m = re.match('^define ([a-zA-Z][a-zA-Z0-9_]*)\s+([a-zA-Z0-9_]*)$', source)
    if m is None:
        return None
    else:
        return m.group(1), m.group(2)


def preprocess(source: List[str]) -> list:
    program_counter = 0
    labels = {}
    destination = []
    for line_num, line in enumerate(source):
        line = line.split(';', 1)[0].strip()
        label, line = split_label(line)
        define = get_define(line)
        if label is not None:
            labels[label] = f'0x{program_counter:03}'
        if define is not None:
            labels[define[0]] = define[1]
            line = ''
        if line:
            program_counter += 1
            destination.append(line)
    for line_num, line in enumerate(destination):
        split = line.split(maxsplit=1)
        mnemonic, operands = split[0], split[1:]
        if operands:
            operands = re.split(',\s*', operands[0])
            for op_num, op in enumerate(operands):
                operands[op_num] = labels.get(op, op)
            destination[line_num] = mnemonic + ' ' + ', '.join(operands)
        else:
            destination[line_num] = mnemonic
    return destination


def assemble(line: str) -> int:
    """
    Assembles a *processed* line of Chip-8 assembly. Processed
    lines should not include any pseudo-instructions, labels,
    or defines.
    :param line:
    :return:
    """
    opcode = line.split(' ', 1)[0]
    formats = get_formats(opcode)
    for format in formats:
        args = parseMnemonic(line, *format['format'])
        if args is not None:
            return format['assemble'](args)
    return None


def assemble_file(source: TextIO, dest: IO[bytes]) -> None:
    lines = preprocess(source.readlines())
    data = []

    for line in lines:
        opcode = assemble(line)
        if opcode is None:
            raise ValueError('Unrecognized line:' + line)
        data.append((opcode & 0xFF00) >> 8)
        data.append(opcode & 0xFF)

    dest.write(bytearray(data))


def main(argv: list) -> None:
    print(argv)
    if len(argv) != 3:
        print('Usage: $ python3 assembler.py <source> <destination>')
        sys.exit(0)
    else:
        _, source_path, dest_path = argv
        with open(source_path, 'r') as source:
            with open(dest_path, 'wb+') as dest:
                assemble_file(source, dest)


if __name__ == '__main__':
    main(sys.argv)

File: src/assembler/assembler.ts
class Assembler {

    private static readonly INSTRUCTIONS: Record<string, [((operand: string) => (string | number))[], ((...v: any[]) => number)][]> = {
        'ADD': [
            [[Assembler.parseI, Assembler.parseV], (_i: any, vx: number) => 0xF01E | (vx << 8)],
            [[Assembler.parseV, Assembler.parseByte], (vx: number, byte: number) => 0x7000 | (vx << 8) | (byte)],
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x8004 | (vx << 8) | (vy << 4)]],
        'AND': [
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x8002 | (vx << 8) | (vy << 4)]],
        'CALL': [
            [[Assembler.parseAddress], (addr: number) => 0x2000 | addr]],
        'CLS': [
            [[], () => 0x00E0]],
        'DRW': [
            [[Assembler.parseV, Assembler.parseV, Assembler.parseNibble], (vx: number, vy: number, n: number) => 0xD000 | (vx << 8) | (vy << 4) | n]],
        'JP': [
            [[Assembler.parseAddress], (addr: number) => 0x1000 | addr],
            [[Assembler.parseV0, Assembler.parseAddress], (_v0: any, addr: number) => 0xB000 | addr]],
        'LD': [
            [[Assembler.parseB, Assembler.parseV], (_b: any, v: number) => 0xF033 | (v << 8)],
            [[Assembler.parseDT, Assembler.parseV], (_dt: any, v: number) => 0xF015 | (v << 8)],
            [[Assembler.parseF, Assembler.parseV], (_f: any, v: number) => 0xF029 | (v << 8)],
            [[Assembler.parseI, Assembler.parseAddress], (_i: any, addr: number) => 0xA000 | addr],
            [[Assembler.parseAI, Assembler.parseV], (_ai: any, v: number) => 0xF055 | (v << 8)],
            [[Assembler.parseST, Assembler.parseV], (_st: any, v: number) => 0xF018 | (v << 8)],
            [[Assembler.parseV, Assembler.parseByte], (v: number, byte: number) => 0x6000 | (v << 8) | byte],
            [[Assembler.parseV, Assembler.parseDT], (v: number, _dt: any) => 0xF007 | (v << 8)],
            [[Assembler.parseV, Assembler.parseAI], (v: number, _ai: any) => 0xF065 | (v << 8)],
            [[Assembler.parseV, Assembler.parseK], (v: number, _k: any) => 0xF00A | (v << 8)],
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x8000 | (vx << 8) | (vy << 4)]],
        'OR': [
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x8001 | (vx << 8) | (vy << 4)]],
        'RET': [
            [[], () => 0x00EE]],
        'RND': [
            [[Assembler.parseV, Assembler.parseByte], (v: number, byte: number) => 0xC000 | (v << 8) | byte]],
        'SE': [
            [[Assembler.parseV, Assembler.parseByte], (v: number, byte: number) => 0x3000 | (v << 8) | byte],
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x5000 | (vx << 8) | (vy << 4)]],
        'SHL': [
            [[Assembler.parseV], (v: number) => 0x800E | (v << 8)],
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x800E | (vx << 8) | (vy << 4)]],
        'SHR': [
            [[Assembler.parseV], (v: number) => 0x8006 | (v << 8)],
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x8006 | (vx << 8) | (vy << 4)]],
        'SKNP': [
            [[Assembler.parseV], (v: number) => 0xE0A1 | (v << 8)]],
        'SKP': [
            [[Assembler.parseV], (v: number) => 0xE09E | (v << 8)]],
        'SNE': [
            [[Assembler.parseV, Assembler.parseByte], (v: number, byte: number) => 0x4000 | (v << 8) | byte],
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x9000 | (vx << 8) | (vy << 4)]],
        'SUB': [
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x8005 | (vx << 8) | (vy << 4)]],
        'SUBN': [
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x8007 | (vx << 8) | (vy << 4)]],
        'SYS': [
            [[Assembler.parseAddress], (addr: number) => addr]],
        'XOR': [
            [[Assembler.parseV, Assembler.parseV], (vx: number, vy: number) => 0x8003 | (vx << 8) | (vy << 4)]],
    };

    public static assembleLines(lines: string[]): number[] | null {
        let result = [];
        for (let line of lines) {
            let opcode = Assembler.assembleLine(line);
            if (opcode === null) {
                return null;
            } else {
                result.push(opcode);
            }
        }
        return result;
    }

    public static assembleLine(line: string): number | null {
        let [mnemonic, operands] = Assembler.splitMnemonic(line);
        return Assembler.toOpcode(mnemonic, operands);
    }

    private static toOpcode(mnemonic: string, operands: string[]): number {
        let formats = Assembler.INSTRUCTIONS[mnemonic];
        if (formats !== undefined) {
            for (let format of formats) {
                let [expectedOperands, conversion] = format;
                let parsedOperands = Assembler.parse(operands, expectedOperands);
                if (parsedOperands !== null && parsedOperands !== undefined) {
                    return conversion(...parsedOperands);
                }
            }
        }
        return null;
    }

    private static parse(operands: string[], parsers: ((s: string) => (string | number))[]): (string | number)[] | null {
        if (operands.length === parsers.length) {
            let ops: (string | number)[] = [];
            for (let i = 0; i < operands.length; i++) {
                ops.push(parsers[i](operands[i]));
                if (ops[i] === null) {
                    return null;
                }
            }
            return ops;
        }
    }


    private static parseV(operand: string): string | number {
        let match = operand.match(/^V([0-9A-F])$/);
        return match !== null ? parseInt(match[1], 16) : null;
    }

    private static parseV0(operand: string): string | number {
        return Assembler.parseLiteral('V0', operand);
    }

    private static parseI(operand: string): string | number {
        return Assembler.parseLiteral('I', operand);
    }

    private static parseB(operand: string): string | number {
        return Assembler.parseLiteral('B', operand);
    }

    private static parseF(operand: string): string | number {
        return Assembler.parseLiteral('F', operand);
    }

    private static parseDT(operand: string): string | number {
        return Assembler.parseLiteral('DT', operand);
    }

    private static parseST(operand: string): string | number {
        return Assembler.parseLiteral('ST', operand);
    }

    private static parseK(operand: string): string | number {
        return Assembler.parseLiteral('K', operand);
    }

    private static parseAI(operand: string): string | number {
        return Assembler.parseLiteral('[I]', operand);
    }

    private static parseNibble(operand: string): string | number {
        let value = parseInt(operand);
        return !isNaN(value) && 0 <= value && value <= 0xF ? value : null;
    }

    private static parseByte(operand: string): string | number {
        let value = parseInt(operand);
        return !isNaN(value) && 0 <= value && value <= 0xFF ? value : null;
    }

    private static parseAddress(operand: string): string | number {
        let value = parseInt(operand);
        return !isNaN(value) && 0 <= value && value <= 0xFFF ? value : null;
    }

    private static parseLiteral(literal: string, operand: string): string | number {
        return operand === literal ? operand : null;
    }

    /**
     * @param line Uppercase, no label, no define, no comment, not empty
     * @returns [opcode, [operands]]
     */
    static splitMnemonic(line: string): [string, string[]] {
        let opEnd = line.indexOf(' ');
        if (opEnd !== -1) {
            let [opcode, operands] = [line.slice(0, opEnd).trim(), line.slice(opEnd).trim()];
            return [opcode, operands.split(/\s*,\s*/).map(o => o.trim())];
        } else {
            return [line, []];
        }
    }
}

try {
    module.exports = Assembler;
} catch (e) {
    // Do nothing
}